### PR TITLE
Fix a bug where the Data Grid cell editor value is dismissed when the user clicks away

### PR DESCRIPTION
**Pull request by joshwooding**
An investigation into an alternative fix to #1462.

This PR reverts https://github.com/jpmorganchase/salt-ds/pull/1462 and tries to solve it differently it avoids API changes and focuses on fixing the bug.

It will be easier to look at specific commits to review it.

The three commits here are:
The revert of the original commit
The commit that adds the fix
A commit that cherrypicks the changeset and tests

### CODE REVIEW

**Review by feliciackh — DISMISSED** [review]

File: packages/grid/src/TextCellEditor.tsx [hunk 1]
```diff
@@ -50,15 +53,26 @@ export function TextCellEditor<T>(props: TextCellEditorProps<T>) {
...
+
+  useEffect(() => {
+    const input = inputRef.current;
+    const focusOut = (event: FocusEvent) => {
+      if (!input?.contains(event.target as Node)) {
+        endEditMode(editorText);
+      }
+    };
+
+    document?.addEventListener("mousedown", focusOut, true);
```
Comment (origami-z) [inline packages/grid/src/TextCellEditor.tsx:70]: Do we need some comment around capture v.s. race condition?

**Review by origami-z — COMMENTED** [review]
I don't think numeric editor works? Was testing on 2nd column below
https://316a45cd.saltdesignsystem-storybook.pages.dev/?path=/story/data-grid-data-grid--editable-cells

File: packages/grid/src/TextCellEditor.tsx (hunk 1 above)
Comment (joshwooding) [reply packages/grid/src/TextCellEditor.tsx:70]: Added :)

**Comment by joshwooding** [comment]
> I don't think numeric editor works? Was testing on 2nd column below https://316a45cd.saltdesignsystem-storybook.pages.dev/?path=/story/data-grid-data-grid--editable-cells

Ah yes, I think it wasn't fixed in the original PR. Just added a commit that fixes this there too. I tested it with Dropdown and it works

File: packages/grid/src/Grid.tsx [hunk 2]
```diff
@@ -471,45 +471,39 @@ export const Grid = function Grid<T>(props: GridProps<T>) {
...
+  const endEditMode = (value: string) => {
+    if (!editMode) {
+      return;
+    }
+    if (cursorColIdx == undefined) {
+      console.error(`endEditMode: cursorColIdx is undefined in edit mode`);
+      return;
+    }
+    const c = cols[cursorColIdx];
+    const handler = c.info.props.onChange;
```
[the remark this answers is not in the document]
Comment (joshwooding) [reply packages/grid/src/Grid.tsx:492]: Those changes are due to a revert and not a manual change. I've made it easier to review the PR by keeping things in separate commits and the first should be ignored.

**Pull request merged**